pkg/component/runtime/service.go

```diff
@@ -0,0 +1,441 @@
+
+	ch       chan ComponentState
+	actionCh chan serviceAction
+	compCh   chan component.Component
+	statusCh chan service.Status
+
+	state          ComponentState
+	lastCheckin    time.Time
+	missedCheckins int
+
+	lastServiceStatus service.Status
+}
+
+// NewServiceRuntime creates a new command runtime for the provided component.
+func NewServiceRuntime(comp component.Component, logger *logger.Logger) (ComponentRuntime, error) {
+	if comp.Spec.Spec.Service == nil {
+		return nil, errors.New("must have service defined in specification")
+	}
+
+	return &ServiceRuntime{
+		current:           comp,
+		log:               logger,
+		ch:                make(chan ComponentState),
+		actionCh:          make(chan serviceAction),
+		compCh:            make(chan component.Component),
+		statusCh:          make(chan service.Status),
+		state:             newComponentState(&comp),
+		lastServiceStatus: service.StatusUnknown,
+	}, nil
+}
+
+// check executes the service check command
+func (s *ServiceRuntime) check(ctx context.Context) error {
+	if s.current.Spec.Spec.Service.Operations.Check == nil {
+		s.log.Errorf("missing check spec for %s service", s.current.Spec.BinaryName)
+		return ErrOperationSpecUndefined
+	}
+	s.log.Debugf("check if the %s is installed", s.current.Spec.BinaryName)
+	return executeServiceCommand(ctx, s.current.Spec.BinaryPath, s.current.Spec.Spec.Service.Operations.Check)
+}
+
+// install executes the service install command
+func (s *ServiceRuntime) install(ctx context.Context) error {
+	if s.current.Spec.Spec.Service.Operations.Install == nil {
+		s.log.Errorf("missing install spec for %s service", s.current.Spec.BinaryName)
+		return ErrOperationSpecUndefined
+	}
+	s.log.Debugf("install %s service", s.current.Spec.BinaryName)
+	return executeServiceCommand(ctx, s.current.Spec.BinaryPath, s.current.Spec.Spec.Service.Operations.Install)
+}
+
+// uninstall executes the service uninstall command
+func (s *ServiceRuntime) uninstall(ctx context.Context) error {
+	if s.current.Spec.Spec.Service.Operations.Uninstall == nil {
+		s.log.Errorf("missing uninstall spec for %s service", s.current.Spec.BinaryName)
+		return ErrOperationSpecUndefined
+	}
+	s.log.Debugf("uninstall %s service", s.current.Spec.BinaryName)
+	return executeServiceCommand(ctx, s.current.Spec.BinaryPath, s.current.Spec.Spec.Service.Operations.Uninstall)
+}
+
+// Run starts the runtime for the component.
+//
+// Called by Manager inside a go-routine. Run should not return until the passed in context is done. Run is always
+// called before any of the other methods in the interface and once the context is done none of those methods will
+// ever be called again.
+// The communicator is currently is not used with the service/daemon component.
+// Perform the basic service status checks until we figure out the comms with the Endpoint.
+func (s *ServiceRuntime) Run(ctx context.Context, comm Communicator) error {
+	s.forceCompState(client.UnitStateStarting, "Starting")
```

Comment:
This should not be set until `Start()` is called.

Reply:
removed